src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationTestFixture.cs

```diff
@@ -83,19 +92,137 @@ public WebApplicationTestFixture(string solutionRelativePath)
+        /// <typeparamref name = "TStartup" /> will be loaded as application assemblies.
+        /// </para>
+        /// </summary>
+        /// <param name="solutionSearchPattern">The glob pattern to use when searching for a solution file by
+        /// traversing up the folder hierarchy from the test execution folder.</param>
+        /// <param name="solutionRelativePath">The path to the project folder relative to the solution file of your
+        /// application. The folder of the first sln file that matches the <paramref name="solutionSearchPattern"/>
+        /// found traversing up the folder hierarchy from the test execution folder is considered as the base path.</param>
+        public static WebApplicationTestFixture<TTestClass, TStartup> Create(
+            string solutionRelativePath,
+            string solutionSearchPattern) =>
+            new WebApplicationTestFixture<TTestClass, TStartup>(solutionRelativePath, solutionSearchPattern);
+
+        private void EnsureServer()
         {
+            if (_server != null)
+            {
+                return;
+            }
+
+            EnsureBuilder();
             EnsureDepsFile();
 
-            var builder = CreateWebHostBuilder();
-            builder
-                .UseStartup<TStartup>()
-                .UseSolutionRelativeContentRoot(solutionRelativePath);
+            SetContentRoot();
 
-            ConfigureWebHost(builder);
-            _server = CreateServer(builder);
+            ConfigureWebHost(_builder);
+            _server = CreateServer(_builder);
+        }
 
-            Client = CreateClient();
+        private void SetContentRoot()
+        {
+            if (_solutionRelativeContentRootPath != null)
+            {
+                if (_solutionSearchPattern == null)
+                {
+                    _builder.UseSolutionRelativeContentRoot(_solutionRelativeContentRootPath);
+                }
+                else
+                {
+                    _builder.UseSolutionRelativeContentRoot(_solutionRelativeContentRootPath, _solutionSearchPattern);
+                }
+            }
+            else
+            {
+                var testAssembly = typeof(TTestClass).Assembly;
+                var metadataAttributes = testAssembly.GetCustomAttributes<AssemblyMetadataAttribute>().ToArray();
+                var startupAssembly = typeof(TStartup).Assembly;
+                var assemblyFile = new Uri(startupAssembly.CodeBase).Segments.Last();
+                var metadataKey = $"Microsoft.AspNetCore.Testing.ContentRoot[{assemblyFile}]";
+                string contentRoot = null;
+                for (var i = 0; i < metadataAttributes.Length; i++)
+                {
+                    var metadataAttribute = metadataAttributes[i];
+                    if (string.Equals(metadataAttribute.Key, metadataKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        contentRoot = metadataAttribute.Value;
```

Comment:
Do you want to break after finding the first match?

Reply:
Yes, I'm still working through the changes. I just pushed so that it gets triggered on travis/appveyor.